src/Components/Components/src/NavigationManager.cs

```diff
@@ -0,0 +1,220 @@
+            }
+            protected set
+            {
+                _uri = value;
+            }
+        }
+
+        /// <summary>
+        /// Navigates to the specified URI.
+        /// </summary>
+        /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI
+        /// (as returned by <see cref="BaseUri"/>).</param>
+        /// <param name="forceLoad">If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
+        public void NavigateTo(string uri, bool forceLoad = false)
+        {
+            AssertInitialized();
+            NavigateToCore(uri, forceLoad);
+        }
+
+        /// <summary>
+        /// Navigates to the specified URI.
+        /// </summary>
+        /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI
+        /// (as returned by <see cref="BaseUri"/>).</param>
+        /// <param name="forceLoad">If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
+        protected abstract void NavigateToCore(string uri, bool forceLoad);
+
+        /// <summary>
+        /// Called to initialize BaseURI and current URI before these values are used for the first time.
+        /// Override <see cref="EnsureInitialized" /> and call this method to dynamically calculate these values.
+        /// </summary>
+        protected void Initialize(string baseUri, string uri)
+        {
+            // Make sure it's possible/safe to call this method from constructors of derived classes.
+            if (uri == null)
+            {
+                throw new ArgumentNullException(nameof(uri));
+            }
+
+            if (baseUri == null)
+            {
+                throw new ArgumentNullException(nameof(baseUri));
+            }
+
+            if (_isInitialized)
+            {
+                throw new InvalidOperationException($"'{typeof(NavigationManager).Name}' already initialized.");
+            }
+            _isInitialized = true;
+
+            Uri = uri;
+            BaseUri = baseUri;
+        }
+
+        /// <summary>
+        /// Allows derived classes to lazyly self-initialize. Implementations that support lazy-initialization should override
+        /// this method and call <see cref="Initialize(string, string)" />.
+        /// </summary>
+        protected virtual void EnsureInitialized()
+        {
+        }
+
+        /// <summary>
+        /// Converts a relative URI into an absolute one (by resolving it
+        /// relative to the current absolute URI).
+        /// </summary>
+        /// <param name="relativeUri">The relative URI.</param>
+        /// <returns>The absolute URI.</returns>
+        public Uri ToAbsoluteUri(string relativeUri)
```

Comment:
I would make ToAbsoluteUri and ToBaseRelativePath protected methods if they are not used outside the implementations.

I'm not sure how beneficial is to expose these functions to external consumers. In fact, I would go as far as to move them to the individual implementations and make them internal.

There are three cases for navigation here, assuming a baseUri of https://www.example.com/base/:
* Absolute URL (external or internal)
  * External -> Follow the link.
  * Internal -> Follow the link, the router will handle it.
* Relative urls "something/else.html" -> https://www.example.com/base/something/else.html -> build full url, follow.
* Host relative urls "/hello/from-host.html" -> "https://www.example.com/hello/from-host.html -> build full url, follow.

NavigateTo can handle these three cases and that way ToAbsoluteUrl and ToBaseRelativePath shouldn't be needed. Same goes for BaseUri as its just an implementation detail.

In fact, the underlying implementations don't even handle this cases and simple forward to the underlying JS concept, which will do the right thing, so could we simply remove these two methods.

ToAbsoluteUri is used only by NavLink, so IMO it can live there as an implementation detail of that class. Implementing this functionality is trivial for users if they need to, but my instinct tells me this is not a common case.

ToBaseRelativePath is only used inside the router, which already gets the BaseURI, so the same reasoning applies there.

We don't have to expose these two methods here as:
* They aren't used in this/underlying implementations.
* They are not used in any code/sample out-side our own components, so they don't support a common use case.
* They are trivial to implement provided you have access to the base Uri, (which you do)

Reply:
Interesting points. If we can do so, I'd be in favour of removing random helper methods from here (now it's been promoted from a *helper* to a *manager*, this class is too busy managing stuff, approving expense reports, etc. to deal with junior tasks like converting URLs).

If `ToAbsoluteUri` can go into `NavLink` and `ToBaseRelativePath` can go into `Router`, that sounds great.

My only reservation would be if we have any evidence that `ToBaseRelativePath` is used widely by customers or is in any samples or docs, but TBH it seems unlikely.

Reply:
I do expect that we'll put things here that help users. If we need utility functionality related to URLs then this is the right place for it.

Reply:
I'll investigate if it's reasonable to move this functionality. We can expose it if people ask.

Reply:
I'm leaving these here. They have somewhat idiosyncratic functionalality that's not easy to duplicate using the APIs on `System.Uri`.